packages/react/src/runtime/ax.ts

```diff
@@ -1,64 +1,74 @@
-const UNDERSCORE_UNICODE = 95;
-
 /**
  * This length includes the underscore,
  * e.g. `"_1s4A"` would be a valid atomic group hash.
  */
 const ATOMIC_GROUP_LENGTH = 5;
 
 /**
- * Joins classes together and ensures atomic declarations of a single group exist.
- * Atomic declarations take the form of `_{group}{value}` (always prefixed with an underscore),
- * where both `group` and `value` are hashes **four characters long**.
- * Class names can be of any length,
- * this function can take both atomic declarations and class names.
- *
- * Input:
- *
- * ```
- * ax(['_aaaabbbb', '_aaaacccc'])
- * ```
+ * Create a classname string which contains only a _single_ classname for each atomic _group_, and that the _last_ atomic definition _value_ remains.
  *
- * Output:
- *
- * ```
+ * ```ts
+ * ax(['_aaaabbbb', '_aaaacccc']);
+ * // output
  * '_aaaacccc'
  * ```
  *
- * @param classes
+ * **Notes**
+ * - Atomic style declarations take the form of `_{group}{value}`, where `group` and `value` are four characters long
+ * - This function will preserve any non atomic style declarations (eg "my-cool-class")
  */
 export default function ax(classNames: (string | undefined | null | false)[]): string | undefined {
-  if (classNames.length <= 1 && (!classNames[0] || classNames[0].indexOf(' ') === -1)) {
-    // short circuit if there's no custom class names.
-    return classNames[0] || undefined;
+  // Shortcut: nothing to do
+  if (!classNames.length) {
+    return;
   }
 
-  const atomicGroups: Record<string, string> = {};
+  // Shortcut: don't need to do anything if we only have a single classname
+  if (
+    classNames.length === 1 &&
+    classNames[0] &&
+    // checking to see if `classNames[0]` is a string that contains other classnames
+    !classNames[0].includes(' ')
```

Comment:
[Support for `String.prototype.includes` is great](https://caniuse.com/es6-string-includes)